examples/components.md

```diff
@@ -0,0 +1,733 @@
+long run, as the number of features and deployments grow. What if you have `N`
+opt-in features and `M` real-world deployment scenarios that ship with `0-N` of
+these features?
+
+Ideally, you want to move each feature under a separate, reusable overlay and
+enable them on-demand per deployment, i.e., in kustomization files of top-level
+overlays. Enter components.
+
+## Components example
+
+Here's an alternative and more [DRY] approach that solves this issue by using a
+Kustomize feature called "components". Each opt-in feature gets packaged as a
+component, so that it can be referred to from higher-level overlays.
+
+First, define a place to work:
+
+```shell
+DEMO_HOME=$(mktemp -d)
+```
+
+Define a common **base** that has a `Deployment` and a simple `ConfigMap`, that
+is mounted on the application's container.
+
+```shell
+BASE=$DEMO_HOME/base
+mkdir $BASE
+
+cat <<EOF >$BASE/kustomization.yaml
+resources:
+- deployment.yaml
+
+configMapGenerator:
+- name: conf
+  literals:
+    - main.conf=|
+        color=cornflower_blue
+        log_level=info
+EOF
+
+cat <<EOF >$BASE/deployment.yaml
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: example
+spec:
+  template:
+    spec:
+      containers:
+      - name: example
+        image: example:1.0
+        volumeMounts:
+        - name: conf
+          mountPath: /etc/config
+      volumes:
+        - name: conf
+          configMap:
+            name: conf
+EOF
+```
+
+Define an `external_db` component, using `kind: Component`, that creates a
+`Secret` for the DB password and a new entry in the `ConfigMap`:
+
+```shell
+EXT_DB=$DEMO_HOME/components/external_db
+mkdir -p $EXT_DB
+
+cat <<EOF >$EXT_DB/kustomization.yaml
+apiVersion: kustomize.config.k8s.io/v1alpha1  # <-- Component notation
+kind: Component
+
+secretGenerator:
+- name: dbpass
+  files:
+    - dbpass.txt
+
+patchesStrategicMerge:
+  - configmap.yaml
+
+patchesJson6902:
+- target:
+    group: apps
+    version: v1
+    kind: Deployment
+    name: example
+  path: deployment.yaml
+EOF
+
+cat <<EOF >$EXT_DB/deployment.yaml
+- op: add
+  path: /spec/template/spec/volumes/0
+  value:
+    name: dbpass
+    secret:
+      secretName: dbpass
+- op: add
+  path: /spec/template/spec/containers/0/volumeMounts/0
+  value:
+    mountPath: /var/run/secrets/db/
+    name: dbpass
+EOF
+
+cat <<EOF >$EXT_DB/configmap.yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: conf
+data:
+  db.conf: |
+    endpoint=127.0.0.1:1234
+    name=app
+    user=admin
+    pass=/var/run/secrets/db/dbpass.txt
+EOF
+```
+
+Define an `ldap` component, that creates a `Secret` for the LDAP password
+and a new entry in the `ConfigMap`:
+
+```shell
+LDAP=$DEMO_HOME/components/ldap
+mkdir -p $LDAP
+
+cat <<EOF >$LDAP/kustomization.yaml
+apiVersion: kustomize.config.k8s.io/v1alpha1
+kind: Component
+
+secretGenerator:
+- name: ldappass
+  files:
+    - ldappass.txt
+
+patchesStrategicMerge:
```

Comment:
Maybe use a configMapGenerator in one of the components to show that it would work as well?  Something like the following (or use a `files:` generator): 
```suggestion
configMapGenerator:
  - name: conf
    behavior: merge
    literals:
      - ldap.conf=|
          endpoint=ldap://ldap.example.com
          bindDN=cn=admin,dc=example,dc=com
          pass=/var/run/secrets/ldap/ldappass.txt
```

Reply:
Sure, let's spice this up.